src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/ProxyHeaderModule.Framework.cs

```diff
@@ -34,12 +34,53 @@ public void Init(HttpApplication context)
 
             context.BeginRequest += (s, e) => UseOptions(((HttpApplication)s).Context.Request);
         }
+
+        context.PreSendRequestHeaders += (s, e) => FixupLocationHeader(((HttpApplication)s).Context);
+    }
+
+    private static void FixupLocationHeader(HttpContext context)
+    {
+        // Some scenarios (such as OIDC auth) will store a URL (including the local host/port rather than the forwarded one)
```

Comment:
where is the OIDC getting the information about host/port? AFAIK part of setting the proxy stuff is so that things were automatically accounted for

Reply:
I haven't dug in to see exactly where it's getting them from, but I found that when a challenge results in redirecting to OIDC login, the URL that the user was trying to navigate to gets stored in a serialized state blob that travels along for the OIDC journey and, at the end, gets spit back out in a Location header to redirect the user to that URL. Somehow the OIDC code that handles challenges was picking up the ASP.NET host/port instead of the ASP.NET Core one even with SERVER_NAME and SERVER_PORT set.

Reply:
Ok, I figured out why this is necessary.

The problematic code is in the OWIN OpenIdConnectAuthenticationHandler: https://github.com/aspnet/AspNetKatana/blob/main/src/Microsoft.Owin.Security.OpenIdConnect/OpenidConnectAuthenticationHandler.cs#L137

When auth is challenged the OWIN OIDC handler creates a redirect URI based on its Request property (of type Microsoft.Owin.OwinRequest). That property has a "Host" member which is not affected by the ServerVariables set on the HttpRequestBase earlier (even though I can see the x-forward headers in the OwinRequest!).

@Tratcher - any ideas how we can fix this on the OWIN side? The summary is that OpenIdConnectAuthenticationHandler is looking at the current URI to figure out where to redirect users back to after auth in challenge scenarios and is coming up with a URI that contains the ASP.NET host/port instead of the ASP.NET Core one even though x-forward headers are set and request ServerVariables like SERVER_NAME and SERVER_PORT have been set. This PR works around the issue by adding some code that watches response headers coming from the ASP.NET app and fixes up any Location headers trying to redirect the user to the old app instead of the new one. Still, if OWIN used the new redirect URI here that workaround wouldn't be necessary.

![image](https://user-images.githubusercontent.com/10077254/177678394-36bd18a4-d34a-4b8b-8ee8-bebffbf706ac.png)

Reply:
I think I found a fix. I was fixing a different bug with the proxy stuff, and we aren't changing the Host header. Owin is using that to grab this info. I've update the module in #92 to handle that (it adds a bunch of tests around that as well, which will slightly conflict with the tests you've added here).

Reply:
Awesome! I'll test that out and if that addresses the issue with OIDC scenarios, then I can revert that part of this PR.

Reply:
It's merged into main now, so hopefully that fixes this issue

Reply:
I've confirmed that https://github.com/dotnet/systemweb-adapters/pull/92 fixes the issue with OIDC not respecting forwarded host/port, so I've reverted this part of the PR.